src/dynamic_honey_badger/dynamic_honey_badger.rs

```diff
@@ -108,6 +215,154 @@ where
+            // Message epoch is out of range.
+            warn!(
+                "{:?}@{} discarded {:?}@{}:{:?}",
+                self.netinfo.our_id(),
+                self.start_epoch,
+                sender_id,
+                start_epoch,
+                content,
+            );
+            return Ok(Fault::new(sender_id.clone(), FaultKind::EpochOutOfRange).into());
+        }
+        let mut step = match content {
+            MessageContent::HoneyBadger(_, hb_msg) => {
+                self.handle_honey_badger_message(sender_id, hb_msg)
+            }
+            MessageContent::KeyGen(_, kg_msg, sig) => self
+                .handle_key_gen_message(sender_id, kg_msg, *sig)
+                .map(FaultLog::into),
+            MessageContent::SignedVote(signed_vote) => self
+                .vote_counter
+                .add_pending_vote(sender_id, signed_vote)
+                .map(FaultLog::into),
+        }?;
+        let our_id = self.netinfo.our_id();
+        let all_remote_validators = self
+            .netinfo
+            .all_ids()
+            .filter(|&id| id != our_id)
+            .into_iter();
+        let all_remote_nodes = all_remote_validators.chain(self.nodes_being_added.iter());
+        let deferred_msgs = step.defer_messages(
+            &self.remote_epochs,
+            all_remote_nodes,
+            self.max_future_epochs as u64,
+        );
+        // Append the deferred messages onto the queues.
+        for (id, message) in deferred_msgs {
+            let epoch = message.dynamic_epoch();
+            if message.is_dynamic() {
+                self.outgoing_queue_dhb
+                    .entry((id, epoch.start_epoch))
+                    .and_modify(|e| e.push(message.clone()))
+                    .or_insert_with(|| vec![message.clone()]);
+            } else {
+                self.outgoing_queue_hb
+                    .entry((id, epoch))
+                    .and_modify(|e| e.push(message.clone()))
+                    .or_insert_with(|| vec![message.clone()]);
+            }
+        }
+        Ok(step)
+    }
+
+    /// Handles an epoch start announcement from a remote `DynamicHoneyBadger`.
+    fn handle_dynamic_epoch_started(
+        &mut self,
+        sender_id: &N,
+        epoch: DynamicEpoch,
+    ) -> Result<Step<C, N>> {
+        debug!(
+            "{:?}@{} received epoch {:?}@{:?}",
+            self.netinfo.our_id(),
+            self.start_epoch,
+            sender_id,
+            epoch,
+        );
+        let mut obsolete = false;
+        self.remote_epochs
+            .entry(sender_id.clone())
+            .and_modify(|e| {
+                if *e < epoch {
+                    // Increment the epoch.
+                    *e = epoch;
+                } else if *e >= epoch {
+                    obsolete = true;
+                }
+            }).or_insert(epoch);
+        if obsolete {
+            return Err(ErrorKind::ObsoleteDynamicEpochStarted.into());
```

Comment:
We don't want to rely on message ordering, so I'd not consider that a fault either.